File: samples/snippets/visualbasic/VS_Snippets_CLR/Asynchronous_File_IO_async/vb/example2.vb
Imports System.IO

Class MainWindow

    ' <snippet2>
    Private Async Sub Button_Click(sender As Object, e As RoutedEventArgs)
        Dim UserDirectory As String = "c:\Users\exampleuser\"

        Using SourceReader As StreamReader = File.OpenText(UserDirectory + "BigFile.txt")
            Using DestinationWriter As StreamWriter = File.CreateText(UserDirectory + "CopiedFile.txt")
                Await CopyFilesAsync(SourceReader, DestinationWriter)
            End Using
        End Using
    End Sub

    Public Async Function CopyFilesAsync(Source As StreamReader, Destination As StreamWriter) As Task
        Dim buffer(4096) As Char
        Dim numRead As Integer

        numRead = Await Source.ReadAsync(buffer, 0, buffer.Length)
        Do While numRead <> 0
            Await Destination.WriteAsync(buffer, 0, numRead)
            numRead = Await Source.ReadAsync(buffer, 0, buffer.Length)
        Loop
        
    End Function
    ' </snippet2>
End Class

File: samples/snippets/csharp/VS_Snippets_CLR/Asynchronous_File_IO_async/cs/example2.cs
using System;
using System.Threading.Tasks;
using System.Windows;
using System.IO;

namespace WpfApplication1
{
    public partial class MainWindow2 : Window
    {
        public MainWindow2()
        {
            InitializeComponent();
        }

	// <snippet2>
        private async void Button_Click(object sender, RoutedEventArgs e)
        {
            string UserDirectory = @"c:\Users\exampleuser\";

            using (StreamReader SourceReader = File.OpenText(UserDirectory + "BigFile.txt"))
            {
                using (StreamWriter DestinationWriter = File.CreateText(UserDirectory + "CopiedFile.txt"))
                {
                    await CopyFilesAsync(SourceReader, DestinationWriter);
                }
            }
        }

        public async Task CopyFilesAsync(StreamReader Source, StreamWriter Destination) 
        { 
            char[] buffer = new char[0x1000]; 
            int numRead; 
            while ((numRead = await Source.ReadAsync(buffer, 0, buffer.Length)) != 0) 
            {
                await Destination.WriteAsync(buffer, 0, numRead);
            } 
        } 
	// </snippet2>
    }
}
